### ISSUE TITLE
Fails to compile `core::mem::swap` that is scceeded previously

### ISSUE THREAD

**Issue by hatoo** [EmbarkStudios/rust-gpu#804]
### Expected Behaviour

Codes that uses `core::mem::swap`

```rust
#![cfg_attr(
    target_arch = "spirv",
    no_std,
    feature(register_attr),
    register_attr(spirv)
)]

#[cfg(not(target_arch = "spirv"))]
use spirv_std::macros::spirv;

use spirv_std as _;

#[spirv(vertex)]
pub fn test_vs() {
    let mut x1 = 0.0f32;
    let mut x2 = 1.0f32;

    core::mem::swap(&mut x1, &mut x2);
}
```

will compile. it's OK in previous version at least b9867d0e84d29d6e60ad62427f453b4829abda56.

### Example & Steps To Reproduce

1. Clone https://github.com/hatoo/rust-gpu-issue/tree/swap-fail (note: swap-fail branch)
2. cargo build and error

```
> cargo build
   Compiling builder v0.1.0 (C:\Users\hato2\Desktop\rust-gpu-issue\builder)
error: failed to run custom build command for `builder v0.1.0 (C:\Users\hato2\Desktop\rust-gpu-issue\builder)`

Caused by:
  process didn't exit successfully: `C:\Users\hato2\Desktop\rust-gpu-issue\target\debug\build\builder-caf084bc583adeed\build-script-build` (exit code: 1)
  --- stderr
     Compiling shader v0.1.0 (C:\Users\hato2\Desktop\rust-gpu-issue\shader)
  error: Cannot memcpy dynamically sized data
      --> C:\Users\hato2\.rustup\toolchains\nightly-2021-10-26-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\intrinsics.rs:2057:14
       |
  2057 |     unsafe { copy_nonoverlapping(src, dst, count) }
       |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  error: could not compile `shader` due to previous error
  Error: BuildFailed
```

It's OK in the previous version. Please see https://github.com/hatoo/rust-gpu-issue/tree/swap-succ.

## System Info

 - Rust: rustc 1.58.0-nightly (29b124802 2021-10-25)
 - OS: Windows 11
 - GPU: RTX2080ti
 - SPIR-V: SPIRV-Tools v2021.3 v2021.3

## Backtrace

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by oisyn** [comment]
Closing as won't fix.

**Comment by eddyb** [comment]
This actually works in Rust-GPU 0.4, and more specifically since `nightly-2022-02-28`, thanks to:
* https://github.com/rust-lang/rust/pull/94412
<sub>(the function being modified there, `swap_simple`, is always used on SPIR-V, and the change that actually unblocked it was `ptr::copy_nonoverlapping(y, x, 1)` being replaced with a `ptr::{read,write}` pair)</sub>

But at the same time, I wanted to propose to upstream that they can remove the `target_arch = "spirv"` hack, as it got brought up in this issue:
* https://github.com/rust-lang/unsafe-code-guidelines/issues/461

~~Now I'm worried that people may be relying on it, since it works, but I'm not sure how to check.~~
~~`core` itself seems to mostly use for sorting/partitioning-like tasks.~~

Oh, [according to sourcegraph, it's definitely gotten used](https://sourcegraph.com/search?q=context:global+language:rust+spirv_std+AND+/swap%5C%28/&patternType=standard&case=yes&sm=1&groupBy=path), oh well.

So we might want to do nothing for now, and only remove the upstream hack if `qptr` can supersede it.
